students/csdotson/lesson05/mailroom_3.py

```diff
@@ -0,0 +1,115 @@
+#!/usr/bin/env python3
+# Lesson 4 - Mailroom... part 2
+
+### Function Definitions ###
+def prompt(prompt, menu):
+    # Create menu based on arguments
+    while True:
+        try:
+            response = input(prompt)
+            if menu[response]() == "quit":
+                break
+        except KeyError:
+            print('Please enter a valid selection!')
+
+
+def send_thank_you():
+    # Create sub_prompt for creating thank you's and adding new donors/donations
+    prompt(thank_you_prompt, thank_you_menu)
+
+
+def add_donation():
+    # Append donation to existing donor, or add donor/donation to existing list
+    try:
+        donor_name = input("\nPlease enter a donor name: ")
+        new_donation = input(f"Please enter a donation amount for {donor_name}: ")
+        if donor_name not in donations:
+            donations[donor_name] = [float(new_donation)]
+        else:
+            donations[donor_name].append(float(new_donation))
+    except ValueError:
+        print("Please enter a numeric value for 'donation'!")
+        add_donation()
+    else:
+        print_email(donor_name, new_donation)
+
+
+def write_letters():
+    # For all donors, generate a thank you letter, and write it to disk as a text file
+    for donor in donations:
+        file_name = '_'.join(donor.split()) + ".txt"
+        with open(file_name, 'w') as f:
+            f.write(f"""Dear {donor},
+
+            Thank you very much for your generosity! Your most recent gift of ${(donations[donor][-1])} will be put to great use. So far, you've donated a total of ${sum(donations[donor])}!
+
+            Sincerely,
+            The Team""")
+
+            print(f"\nCreated letter for {donor}!")
+
+
+def print_email(donor, amount):
+    # Print formatted email thanking donor
+    letter_details = {'name': donor, 'donation_amount': amount}
+    letter = "\nDear {name},\n\nThank you so very much for your kind donation of ${donation_amount}. We can assure you that it will be put to great use.\n\nBest,\nChris".format(**letter_details)
+    print(letter)
+
+
+def create_report():
+    # Print formatted report showing donors and donation stats
+    header = '{:20}|{:^15}|{:^15}|{:>15}'.format("Donor Name", "Total Given", "Num Gifts", "Average Gift")
+    print()
+    print(header)
+    print("-" * len(header))
+
+    for donor, donor_data in donations.items():
+        total_given = sum(donor_data)
+        num_gifts = len(donor_data)
+        avg_gift = total_given / num_gifts
+        print('{:21}{:>15.2f}{:>16}{:>16.2f}'.format(donor, total_given, num_gifts, avg_gift))
+    print()
+
+
+def list_donors():
+    # Formatted print of donor names
+    print("\nList of donors:")
+    [print(donor) for donor in donations]
+    print()
```

Comment:
(see my post in slack on this)

this is better as a for loop, or have your function return a string that can be printed (which can be utilized once u get to unit testing):
```
def list_donors():
    return "\n".join(list(donations))
```